test/unit/org/apache/cassandra/db/filter/ColumnFilterTest.java

```diff
@@ -18,53 +18,348 @@
 
-        testRoundTrip(ColumnFilter.selection(metadata, metadata.partitionColumns().without(v1)), metadata, MessagingService.VERSION_30);
-        testRoundTrip(ColumnFilter.selection(metadata, metadata.partitionColumns().without(v1)), metadata, MessagingService.VERSION_3014);
+    @Test
+    public void testSelectIndividualCellsFromStatic() throws Exception
+    {
+        ColumnFilter filter = ColumnFilter.selectionBuilder().select(s2, path1).select(s2, path3).build();
+        testRoundTrips(filter);
+        assertEquals("[s2[1], s2[3]]", filter.toString());
+        assertFetchedQueried(true, true, filter, s2);
+        assertFetchedQueried(false, false, filter, v1, v2, s1);
+        assertCellFetchedQueried(false, false, filter, v2, path0, path1, path2, path3, path4);
+        assertCellFetchedQueried(true, true, filter, s2, path1, path3);
+        assertCellFetchedQueried(false, false, filter, s2, path0, path2, path4);
     }
 
-    static void testRoundTrip(ColumnFilter columnFilter, CFMetaData metadata, int version) throws Exception
+    @Test
+    public void testSelectCellSlice() throws Exception
     {
-        DataOutputBuffer output = new DataOutputBuffer();
-        serializer.serialize(columnFilter, output, version);
-        Assert.assertEquals(serializer.serializedSize(columnFilter, version), output.position());
-        DataInputPlus input = new DataInputBuffer(output.buffer(), false);
-        Assert.assertEquals(serializer.deserialize(input, version, metadata), columnFilter);
+        ColumnFilter filter = ColumnFilter.selectionBuilder().slice(v2, path1, path3).build();
+        testRoundTrips(filter);
+        assertEquals("[v2[1:3]]", filter.toString());
+        assertFetchedQueried(true, true, filter, v2);
+        assertFetchedQueried(false, false, filter, v1, s1, s2);
+        assertCellFetchedQueried(true, true, filter, v2, path1, path2, path3);
+        assertCellFetchedQueried(false, false, filter, v2, path0, path4);
+        assertCellFetchedQueried(false, false, filter, s2, path0, path1, path2, path3, path4);
+    }
+
+    @Test
+    public void testSelectCellSliceFromStatic() throws Exception
+    {
+        ColumnFilter filter = ColumnFilter.selectionBuilder().slice(s2, path1, path3).build();
+        testRoundTrips(filter);
+        assertEquals("[s2[1:3]]", filter.toString());
+        assertFetchedQueried(true, true, filter, s2);
+        assertFetchedQueried(false, false, filter, v1, v2, s1);
+        assertCellFetchedQueried(false, false, filter, v2, path0, path1, path2, path3, path4);
+        assertCellFetchedQueried(true, true, filter, s2, path1, path2, path3);
+        assertCellFetchedQueried(false, false, filter, s2, path0, path4);
+    }
+
+    @Test
+    public void testSelectColumnsWithCellsAndSlices() throws Exception
+    {
+        ColumnFilter filter = ColumnFilter.selectionBuilder().add(v1).add(s1).slice(v2, path0, path2).select(v2, path4).select(s2, path0).slice(s2, path2, path4).build();
```

Comment:
Nit: we could break this long line:
```suggestion
        ColumnFilter filter = ColumnFilter.selectionBuilder()
                                          .add(v1)
                                          .add(s1)
                                          .slice(v2, path0, path2)
                                          .select(v2, path4)
                                          .select(s2, path0)
                                          .slice(s2, path2, path4)
                                          .build();
```